### PR TITLE
Version 1.0.0

### PR DESCRIPTION
**Pull request by NIKHIL0VERMA**
Created the Chrome-Extension for PickPalette

### CODE REVIEW

**Comment by ArnabChatterjee20k** [comment]
Provide screenshots and a video showing the working

**Review by ArnabChatterjee20k — COMMENTED** [review]
What is this file for??

**Comment by ArnabChatterjee20k** [comment]
Just a query, can't we use npm and vite while developing chrome extensions??

**Comment by NIKHIL0VERMA** [comment]
> Provide images, screenshots and a deployed link

Okay, will do tomorrow at 9 pm due to clg and commute. The zip file provide in issue contains only the chrome extension which can be installed.

Download and Extract the zip.
Enable extension developer option Google it
Open the chrome go to chrome://extension
Drag and drop the crx file provided in zip in chrome. 
Open any website. 
Click the extension click the button in popup. 
Click any colour to copy hex code from palette.

Did you mean I should publish this to chrome web store??

**Comment by NIKHIL0VERMA** [comment]
> What is this file for??

Which one??

**Comment by NIKHIL0VERMA** [comment]
> Just a query, can't we use npm and vite while developing chrome extensions??

We can use react and npm with performance trade off. 🥹
I think we should go with native plain HTML css and JS as its fast.

**Comment by ArnabChatterjee20k** [comment]
> > Provide images, screenshots and a deployed link
> 
> Okay, will do tomorrow at 9 pm due to clg and commute. The zip file provide in issue contains only the chrome extension which can be installed.
> 
> Download and Extract the zip. Enable extension developer option Google it Open the chrome go to chrome://extension Drag and drop the crx file provided in zip in chrome. Open any website. Click the extension click the button in popup. Click any colour to copy hex code from palette.
> 
> Did you mean I should publish this to chrome web store??

No need to share the link. I edited the message. Just share a video whenever you are free.
Thank you for your efforts.

**Comment by ArnabChatterjee20k** [comment]
> > Just a query, can't we use npm and vite while developing chrome extensions??
> 
> We can use react and npm with performance trade off. 🥹 I think we should go with native plain HTML css and JS as its fast.

Ok make sense. But can't we even install packages with npm that I asked? Like if I wanna setup a npm project with vanilla js and do some npm installs and then build and push it.
Its not a great issue but developer experience otherwise here we have to either use cdns or manually script pasting

**Comment by ArnabChatterjee20k** [comment]
> > What is this file for??
> 
> Which one??

The .pem and .crx files. Are they playing any major role? Cause I saw they are having some private keys

**Comment by NIKHIL0VERMA** [comment]
> The .pem and .crx files. Are they playing any major role? Cause I saw they are having some private keys

Oh, don't worry .pem is just the previous private key used to sign the Chrome extension.
.crx is the main Chrome extension file. This means like .exe or .out file is generated when compiling the c, cpp or similar codebase.

>Ok make sense. But can't we even install packages with npm that I asked? Like if I wanna setup a npm project with vanilla js and do some npm installs and then build and push it.
Its not a great issue but developer experience otherwise here we have to either use cdns or manually script pasting

Never tried but I think it's possible. Can I know if you have any particular need for the npm module? what feature do you want to implement?

> Just share a video whenever you are free.

Sure, here we go ;)

https://github.com/ArnabChatterjee20k/PickPalette-ChromeExtension/assets/49404370/18bbb067-34b8-4eb5-a528-e7b4b98f1922

**Comment by ArnabChatterjee20k** [comment]
This is working really good. I am really tempted to merge this as soon as possible. 
Any plans of kinda automating it. Like the deployment process. Like in the main repo whenever I will merge anything to master or main, the netlify will start an automatic build.
Any such process present in chrome extensions

**Comment by NIKHIL0VERMA** [comment]
Netlify is for web, but we can use Github actions. I searched and found this https://medium.com/@ganeshjaiwal/automate-chrome-extension-publish-using-an-automated-script-3b3db459e901. I almost completed the automation part. Things needed to complete automation from your side.

- [x] Change author email in manifest.json
- [ ] Build locally and publish to chrome web store.
- [x] Use Github secret and tweak deploy.js to include your secrets. TODO marked in files.

Things to remember
Don't push to main too frequently use seperate stage branch
Publish Chrome extension with your account before using the automation
Best Regards,
Nikhil Verma

File: deploy.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+jsonfile.writeFileSync(path.join(buildLocation, "manifest.json"), manifest);
+
+// create zip
+zipper.sync.zip(buildLocation).compress().save(path.join(buildLocation, "build.zip"));
+const fileBin = fs.readFileSync(path.join(buildLocation, "build.zip"));
+
+// TODO change the client id and clientSecret
+const chromeWebstore = new ChromeWebstore("<CLIENT_ID>", "<CLIENT_SECRET>");
+// Get new token with refresh_token
+chromeWebstore.getRefreshToken("<REFRESH_TOKEN>").then(function (data) {
```
Comment (ArnabChatterjee20k) [inline deploy.js]: take the refresh token from ```process.env```

**Comment by ArnabChatterjee20k** [comment]
I will review it by tomorrow morning
Till then you can work on other issues as well if you want

**Comment by ArnabChatterjee20k** [comment]
@NIKHIL0VERMA 
Its working great.

File: Web2palette/manifest.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        },
+        "default_title": "Web2Palette",
+        "default_popup": "pages/popup.html"
+    },
+    "icons": {
+        "16": "./imgs/icon16.png",
+        "48": "./imgs/icon48.png",
+        "128": "./imgs/icon128.png"
+    },
+    "author": "nikhil2003verma@gmail.com",
```
Comment (ArnabChatterjee20k) [inline Web2palette/manifest.json]: change the author to pickapalette

File: deploy.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+manifest.version = version;
+// save manifest file
+jsonfile.writeFileSync(path.join(buildLocation, "manifest.json"), manifest);
+
+// create zip
+zipper.sync.zip(buildLocation).compress().save(path.join(buildLocation, "build.zip"));
+const fileBin = fs.readFileSync(path.join(buildLocation, "build.zip"));
+
+// TODO change the client id and clientSecret
+const chromeWebstore = new ChromeWebstore("<CLIENT_ID>", "<CLIENT_SECRET>");
```
Comment (ArnabChatterjee20k) [inline deploy.js]: use process.env to get variables from .env instead of strings

File: package.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "pickpalette-chromeextension",
+  "version": "1.0.0",
+  "description": "Grab color palette from current tab",
+  "scripts": {
+    "deploy": "node deploy.js"
+  },
+  "repository": {
```
Comment (ArnabChatterjee20k) [inline package.json]: remove the repository here

File: package.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    "url": "git+https://github.com/NIKHIL0VERMA/PickPalette-ChromeExtension.git"
+  },
+  "keywords": [
+    "Chrome-extension",
+    "color picker",
+    "palette generator",
+    "scrap colors"
+  ],
+  "author": "Nikhil Verma",
+  "bugs": {
```
Comment (ArnabChatterjee20k) [inline package.json]: remove the bugs section here

File: package.json [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    "url": "https://github.com/NIKHIL0VERMA/PickPalette-ChromeExtension/issues"
+  },
+  "homepage": "https://github.com/NIKHIL0VERMA/PickPalette-ChromeExtension#readme",
```
Comment (ArnabChatterjee20k) [inline package.json]: remove the home page

File: deploy.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+const zipper = require("zip-local");
+const jsonfile = require('jsonfile');
+const path = require('path');
+const ChromeWebstore = require('chrome-webstore-manager');
+
+const itemId = "EXTENSION_ID"; // TODO Change this after publishing the extension on web store
+var buildLocation = path.join(__dirname, "Web2palette");
+
+// read manifest file
+var manifest = jsonfile.readFileSync(path.join(buildLocation, "manifest.json"));
```
Comment (ArnabChatterjee20k) [inline deploy.js]: don't use var
use const here

File: deploy.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+function getNewVersion() {
+    var ver = parseInt(manifest.version.split('.')[0]);
```
Comment (ArnabChatterjee20k) [inline deploy.js]: dont use var

File: deploy.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    ver++;
+    return `${ver}.0.0`;
+}
+var version = getNewVersion();
```
Comment (ArnabChatterjee20k) [inline deploy.js]: dont use var

**Review by ArnabChatterjee20k — CHANGES_REQUESTED** [review]

File: Web2palette/manifest.json (hunk 2 above)
Comment (NIKHIL0VERMA) [reply Web2palette/manifest.json]: Author: it should contain the mail address. It should be the same as the mail used to login to the chrome web store for extension publishing.

Comment (ArnabChatterjee20k) [reply Web2palette/manifest.json]: then I think making it as a env variable will be best. Or Dont commit the manifest file
Write in the readme a sample manifest file and users will modify it according to the choice

**Review by ArnabChatterjee20k — APPROVED** [review]
Fine, I am merging it

**Pull request merged**